feat(yugabytedb): Add YugabyteDB Source and Tool

Sfurti-yb opened the pull request:
- This PR aims to add YugabyteDB as a source and a tool.
- It is based on the PostgreSQL source but uses the YugabyteDB's fork of pgx driver that better accommodates the distributed nature of YugabyteDB
- Added tests for the same.

Sfurti-yb commented:
Hi @kurtisvg apologies for the delay. I am just checking with the team internally if it needs to be signed as an individual or on behalf of YugabyteDB. Will get back to you by today. Thanks.

kurtisvg commented:
@Sfurti-yb Sounds good!

duwenxin99 commented on internal/sources/yugabytedb/yugabytedb.go:
```diff
@@ -0,0 +1,127 @@
+	User                            string `yaml:"user" validate:"required"`
+	Password                        string `yaml:"password" validate:"required"`
+	Database                        string `yaml:"database" validate:"required"`
+	LoadBalance                     string `yaml:"load_balance"`                     // optional
```
1. You could delete the `// optional` comments. The `required` tag is sufficient.
2. The yaml variable names should use camel case instead of snake case.
Thanks!

duwenxin99 commented on docs/en/resources/sources/yugabytedb.md:
```diff
@@ -0,0 +1,44 @@
+| load_balance                      |  boolean  |     false     | If true, enable uniform load balancing.                    |
+| topology_keys                     |  string  |     false     | Comma-separated geo-locations in the form cloud.region.zone:priority to enable topology-aware load balancing. Ignored if load_balance is false.                    |
+| yb_servers_refresh_interval       |  integer  |     false     | The interval (in seconds) to refresh the servers list; ignored if load_balance is false                    |
+| fallback_to_topology_keys_only    |  boolean  |     false     | If set to true and topology_keys are specified, only connect to nodes specified in topology_keys                    |
```
Can we add default values here in the descriptions? Thanks!

duwenxin99 commented:
@Sfurti-yb Hi! I'm trying to set up a Yugabyte DB instance on a GCE VM but I'm running into problems with the server always stuck in the "bootstrapping" status. Would you help me to understand the error here? For context: 
1. I'm using Ubuntu 22.04.5 LTS
2. The commands I'm running is exactly what's on the official X86 installation guide:
```wget https://software.yugabyte.com/releases/2024.2.3.2/yugabyte-2024.2.3.2-b6-linux-x86_64.tar.gz
tar xvfz yugabyte-2024.2.3.2-b6-linux-x86_64.tar.gz && cd yugabyte-2024.2.3.2/
./bin/post_install.sh
./bin/yugabyted start```
3. The error logs are attached below:
[yugabyted-2025-07-07-18h47m35.853223s.tar.gz](https://github.com/user-attachments/files/21108854/yugabyted-2025-07-07-18h47m35.853223s.tar.gz)
Thank you!

Sfurti-yb commented:
@kurtisvg We have signed the enterprise CLA but still the CLA check is failing.
@duwenxin99 We are looking into the error, will get back to you in some time

kurtisvg commented:
@Sfurti-yb I refreshed the check, looks like it's passing now

duwenxin99 commented:
> @duwenxin99 Can you please paste the error message you saw on the terminal? Also, can you let us know the spec of the VM (cpu/memory) used? Could not reproduce this issue with the same Ubuntu/YugabyteDB versions.

Hi @ashetkar, I actually did not see any error messages on the terminal. It started successfully but the status just kept saying "bootstrapping". I just checked again and now it says "stopped".
<img width="1370" height="464" alt="Screenshot 2025-07-10 at 7 56 42 PM" src="https://github.com/user-attachments/assets/67af9e6e-feb9-423e-aa2a-e664323891d4" />
<img width="738" height="173" alt="Screenshot 2025-07-10 at 7 56 59 PM" src="https://github.com/user-attachments/assets/2a6cccfa-af72-4d05-b917-bc5c27f1a024" />
I'm using a 2-vCPU, 8-GB RAM VM:
<img width="482" height="203" alt="Screenshot 2025-07-10 at 8 00 30 PM" src="https://github.com/user-attachments/assets/56875bd9-7b4d-467f-993b-128fcbb4168c" />

ashetkar commented:
@duwenxin99 Apologies for the delay. I'm still unable to reproduce the issue even on a 1-vCPU VM.
Did you check the status via `./bin/yugabyted status` command after the `start` command returned status as `Bootstrapping`?
Generally, in such cases, the cluster shows the status as `Running` after a few (10-20) seconds.
I've reached out to our internal team with your logs.

ashetkar commented:
@duwenxin99 Let me know if we can connect over a call to get you unblocked on this. You can reach out to me at ashetkar@yugabyte.com

ashetkar commented:
@duwenxin99 A gentle reminder. Can we connect to get you unblocked on this?

duwenxin99 commented:
Hi @ashetkar, these are the logs from the recent server run:
- [master_logs_part1.tar.gz](https://github.com/user-attachments/files/21640881/master_logs_part1.tar.gz)
- [master_logs_part2.tar.gz](https://github.com/user-attachments/files/21640884/master_logs_part2.tar.gz)
- [tserver_logs.tar.gz](https://github.com/user-attachments/files/21640888/tserver_logs.tar.gz)
This may be missing some of the info entries because the `collect_log` command wasn't finished after 40min so I terminated it early.

duwenxin99 commented:
@ashetkar @Sfurti-yb 
I don't have permission to your forked repo. Could you update the integration.cloudbuild.yaml file to include:
```
- id: "yugabyte"
    name : golang:1
    waitFor: ["compile-test-binary"]
    entrypoint: /bin/bash
    env:
      - "GOPATH=/gopath"
      - "YUGABYTE_DATABASE=$_YUGABYTE_DATABASE"
      - "SERVICE_ACCOUNT_EMAIL=$SERVICE_ACCOUNT_EMAIL"
    secretEnv: ["YUGABYTE_USER", "YUGABYTE_PASS", "CLIENT_ID"]
    volumes:
      - name: "go"
        path: "/gopath"
    args:
      - -c
      - |
        ./valkey.test -test.v
```
And add these in the `substitution` section:
```
_YUGABYTE_DATABASE: "yugabyte"
_YUGABYTE_PORT: "5433"
```

Sfurti-yb commented:
@duwenxin99 I’ve added the above code and made a small adjustment to this section:

```yaml
args:
  - -c
  - |
    ./valkey.test -test.v
```

changed to:

```yaml
args:
  - -c
  - |
    ./yugabytedb.test -test.v
```

I thought this might be the correct command for our setup—could you please confirm?

duwenxin99 commented:
/gcbrun

duwenxin99 commented:
> @duwenxin99 I’ve added the above code and made a small adjustment to this section:
> 
> ```yaml
> args:
>   - -c
>   - |
>     ./valkey.test -test.v
> ```
> 
> changed to:
> 
> ```yaml
> args:
>   - -c
>   - |
>     ./yugabytedb.test -test.v
> ```
> 
> I thought this might be the correct command for our setup—could you please confirm?

Thanks @Sfurti-yb! That's correct. Could you also add these to the secrets section:
```
- versionName: projects/$PROJECT_ID/secrets/yugabyte_host/versions/latest
      env: YUGABYTE_HOST
    - versionName: projects/$PROJECT_ID/secrets/yugabyte_user/versions/latest
      env: YUGABYTE_USER
    - versionName: projects/$PROJECT_ID/secrets/yugabyte_pass/versions/latest
      env: YUGABYTE_PASS
```

Sfurti-yb commented:
@duwenxin99 I have updated the naming pattern to `YUGABYTEDB_*` and also rebased the branch with main

duwenxin99 commented:
/gcbrun

duwenxin99 commented:
@Sfurti-yb could you give me permission to update to your forked repo? I think this would make it easier to set the test configs. Otherwise please add "YUGABYTEDB_PORT=$_YUGABYTEDB_PORT" and "YUGABYTEDB_LOADBALANCE =$_YUGABYTEDB_LOADBALANCE" to the config "env" section

Sfurti-yb commented:
@duwenxin99 I have updated it now.  Please take a look

duwenxin99 commented:
/gcbrun

ashetkar commented:
Hi @duwenxin99, the branch is rebased now, please proceed as needed. Thanks!

duwenxin99 commented on internal/tools/yugabytedbsql/yugabytedbsql.go:
```diff
@@ -0,0 +1,174 @@
+// Copyright 2024 Google LLC
```
```suggestion
// Copyright 2025 Google LLC
```

duwenxin99 commented on tests/yugabytedb/yugabytedb_integration_test.go:
```diff
@@ -0,0 +1,159 @@
+// Copyright 2024 Google LLC
```
```suggestion
// Copyright 2025 Google LLC
```

duwenxin99 commented on internal/sources/yugabytedb/yugabytedb.go:
```diff
@@ -0,0 +1,127 @@
+// Copyright 2024 Google LLC
```
```suggestion
// Copyright 2025 Google LLC
```

duwenxin99 commented on internal/sources/yugabytedb/yugabytedb_test.go:
```diff
@@ -0,0 +1,285 @@
+package yugabytedb_test
```
```suggestion
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package yugabytedb_test
```

duwenxin99 commented on internal/tools/yugabytedbsql/yugabytedbsql_test.go:
```diff
@@ -0,0 +1,214 @@
+// Copyright 2024 Google LLC
```
```suggestion
// Copyright 2025 Google LLC
```

duwenxin99 commented on internal/tools/yugabytedbsql/yugabytedbsql.go:
```diff
@@ -0,0 +1,174 @@
+	mcpManifest tools.McpManifest
+}
+
+func (t Tool) Invoke(ctx context.Context, params tools.ParamValues) (any, error) {
```
```suggestion
func (t Tool) Invoke(context.Context, tools.ParamValues, tools.AccessToken) (any, error) {
```

duwenxin99 commented on internal/tools/yugabytedbsql/yugabytedbsql.go:
```diff
@@ -0,0 +1,174 @@
+
+func (t Tool) Authorized(verifiedAuthServices []string) bool {
+	return tools.IsAuthorized(t.AuthRequired, verifiedAuthServices)
+}
```
```suggestion
}

func (t Tool) RequiresClientAuthorization() bool {
	return t.UseClientOAuth
}
```

duwenxin99 replied:
Nice catch. Returning false is correct here.

duwenxin99 reviewed (changes requested):
@
